### ISSUE TITLE
Possible bug with running functions in NUMA (multi CPU) environment

### ISSUE THREAD

**Issue by Hellisotherpeople** [openfaas/faas-netes#334]
I have a function - say, a function to compute prime numbers. I am doing tests to see how well these functions are able to stay isolated to the DRAM which is specifically pinned to only a single CPU of a multi CPU system using numactl --membind=0 --cpunodebind=0 on a python3-debian function template

**Expected Behaviour**
Function respects the numactl --membind=0 --cpunodebind=0 settings, and doesn't use the CPU or memory pinned for it on Socket 1

**Current Behaviour**
Functions appear to utilize both socket 0 and socket 1 DRAM - no matter what we do.

**Possible Solution**
We apply the numactl change into the "python3-debian" template Dockerfile, as so:
ENV fprocess="numactl --membind=0 --cpunodebind=0 python3 index.py"
Maybe this is the wrong location, or method, for trying to force each function using this template to respect our numa settings

**Steps to Reproduce (for bugs)**
We've observed this with any function, so the function doesn't matter

have numa system with at least 2 sockets
apply numactl membinding and cpubinding to the template docker file
use a memory measuring tool to monitor the per-socket DRAM usage
Your Environment
Fedora 27,

**FaaS-CLI version ( Full output from: faas-cli version ):**
0.7.5, commit: dcddef2ae92880f50444103963408451abe82cc4

**Docker version docker version (e.g. Docker 17.0.05 ):**
Docker Client + Server are 18.03.1-ce

**Are you using Docker Swarm or Kubernetes (FaaS-netes)?**
Kubernetes (Using helm)

Note, this issue is also tracked on the OpenFaaS proper issue tracker: https://github.com/openfaas/faas/issues/994

**Comment by alexellis** [comment]
Hi @Hellisotherpeople, I'm really not sure how we can engage with your issue with NUMA. Can you help me understand how we could help?

Are you sure this is even anything to do with OpenFaaS and not just an artifact of using a container?

Some ideas: use of-watchdog instead and the HTTP mode. Try your test on your host directly with the watchdog and no containers involved. 

Is your function container privileged enough to send these parameters to their destination? CPU/firmware/kernel etc. 

Alex

**Comment by Hellisotherpeople** [comment]
Hi @alexellis 

It is possible that this is an artifact about using a container - and we are leaning on it being something to do with Docker and not you guys - but I must make sure.

We tried using of-watchdog - though we've found that it's not behaving as a drop-in replacement (possibly due to user error - will investigate further), including with and without HTTP mode. 

For this project, we must use containers - and we're using sudo and attempting to priviledge the containers / minikube / kubernetes as much as possible so as to allow for these numabinding settings to be propegated.
